feat: add keyword validate

polvalente opened the pull request:
This PR implements `Keyword.validate/2` and `Keyword.validate!/2` as discussed in the mailing list.

The original implementation for `validate!/2` can be found in [Nx.Defn.Kernel](https://github.com/elixir-nx/nx/blob/c8353de695a3d70dc6d518e0ab2ec832faa7df68/nx/lib/nx/defn/kernel.ex#L741)

polvalente commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,193 @@ defmodule Keyword do
   end
 
+  @doc """
+  Ensures the first argument is a `keyword` with the given
+  keys and default values.
+  The second argument must be a list of atoms, specifying
+  a given key, or tuples specifying a key and a default value.
+  If any of the keys in the `keyword` is not defined on
+  `values`, it raises an error.
```
This doc is almost the same as the one for `validate/2`. Is there a best-practices approach to avoid being this repetitive, or is the way this was done fine?

wojtekmach replied:
perhaps something like this?

```
  @doc """
  Similar to `validate/2` but returns the keyword or raises an error.

  ## Examples

      iex> Keyword.validate!([two: 3], [one: 1, two: 2])
      [one: 1, two: 3]

      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]

  """
```

btw,  I'd define the `validate!/2` function _after_ `validate/2` given they'd show up in docs in that order

polvalente commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,193 @@ defmodule Keyword do
+  Possible return values:
+  * `{:ok, keyword}`: Success case. Returns the validated keyword with default values applied.
+  * `{:error, :not_a_keyword}`: Indicates that the validated value is not a keyword.
+  * `{:error, :invalid_expected_values}`: Indicates that the expected values list is invalid.
```
I'm not happy with some of these error messages, especially "invalid_expected_values". Suggestions are welcome!

wojtekmach replied:
perhaps something like this?

```
  @doc """
  Similar to `validate/2` but returns the keyword or raises an error.

  ## Examples

      iex> Keyword.validate!([two: 3], [one: 1, two: 2])
      [one: 1, two: 3]

      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]

  """
```

btw,  I'd define the `validate!/2` function _after_ `validate/2` given they'd show up in docs in that order

wojtekmach replied:
for these particular cases, instead of returning an error tuple I'd raise. This would be consistent with e.g. Keyword.fetch:

```
iex(1)> Keyword.fetch!("not a keyword", :key)
** (FunctionClauseError) no function clause matching in Keyword.fetch!/2

    The following arguments were given to Keyword.fetch!/2:

        # 1
        "not a keyword"

        # 2
        :key

    Attempted function clauses (showing 1 out of 1):

        def fetch!(keywords, key) when is_list(keywords) and is_atom(key)

    (elixir 1.13.0-dev) lib/keyword.ex:417: Keyword.fetch!/2
```

```
iex(2)> Keyword.fetch!([], "bad key")
** (FunctionClauseError) no function clause matching in Keyword.fetch!/2

    The following arguments were given to Keyword.fetch!/2:

        # 1
        []

        # 2
        "bad key"

    Attempted function clauses (showing 1 out of 1):

        def fetch!(keywords, key) when is_list(keywords) and is_atom(key)

    (elixir 1.13.0-dev) lib/keyword.ex:417: Keyword.fetch!/2
```

wdyt?

polvalente replied:
Makes sense. I thought that maybe accepting invalid inputs would be desirable as well, but this makes for a better interface

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,193 @@ defmodule Keyword do
+
+      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
+      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]
+  """
```
```suggestion
  """
  @doc since: "1.13.0"
```

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,193 @@ defmodule Keyword do
+      {:error, :invalid_expected_values}
+  """
+  @spec validate(keyword :: keyword(), values :: [atom() | {atom(), term()}]) ::
+          {:ok, keyword}
+          | {:error,
+             :not_a_keyword
+             | {:invalid_key, term()}
+             | :invalid_expected_values}
```
per the ML thread:

```suggestion
          {:ok, keyword} | {:error, unknown_keys, missing_keys}
```

josevalim replied:
To add to @wojtekmach's comment, invalid inputs (i.e. not a keyword or invalid expected values) should just raise.

josevalim replied:
Actually, my comment on the mailing list was wrong, there is no such thing as missing keys as we don't have required keys. So this should simply be:

    {:ok, keyword} | {:error, unknown_keys :: [atom]}

wojtekmach replied:
Oh, good call. In that case, `validate!` should definitely raise a KeyError which will give us a really good error message.

wojtekmach replied:
scratch that, we can't raise a KeyError as we could have multiple unknown keys.

polvalente replied:
@josevalim I've updated the code with the reviews. One thing that isn't clear to me is that the suggested spec shows a list of invalid keys.

The current implementation only returns the first invalid key. Should I change this for `validate`?

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,182 @@ defmodule Keyword do
+  @doc since: "1.13.0"
+  @spec validate(keyword :: keyword(), values :: [atom() | {atom(), term()}]) ::
+          keyword | no_return
+  def validate!(keyword, values) when is_list(keyword) and is_list(values) do
+    # We use two lists to avoid reversing/concatenating
+    # lists in the middle of traversals.
+    case validate!(keyword, values, [], []) do
+      {:ok, keyword} ->
+        keyword
+
+      error ->
+        keys =
+          for value <- values do
+            if is_atom(value) do
+              value
+            else
+              elem(value, 0)
+            end
+          end
+
+        case error do
+          {:badkey, key} ->
+            raise ArgumentError,
+                  "unknown key #{inspect(key)} in #{inspect(keyword)}, " <>
+                    "expected one of #{inspect(keys)}"
+
+          :badkey ->
+            raise ArgumentError,
+                  "expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"
+        end
+    end
+  end
```
ideally the raising variant uses the non-raising variant under the hood, I think it will remove a fair amount of duplication. We can even drop the guards as the underlying function has them already.

```elixir
  def validate!(keyword, values) do
    case validate(keyword, values) do
      {:ok, keyword} ->
        keyword

      {:error, missing_keys} ->
        # ...
    end
  end
```

regarding error, I think we have two options, perhaps something along the lines of either:

1. `"expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"`
2. `"unknown keys #{inspect(missing_keys)} in #{inspect(keyword)}, the allowed keys are: #{inspect(keys)}"`

what do you think?

polvalente replied:
Yeah, I'll refactor the implementation then. Makes more sense, especially since both fail exactly on the same cases.

One thing I noticed is that although probably not intended `KeyError` accepts a list as `:key`. The message in itself becomes a little bit weird though, so I'll stick to ArgumentError

```elixir
iex(1)> raise KeyError, key: [:a, :b, :c]
** (KeyError) key [:a, :b, :c] not found
```

polvalente commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,182 @@ defmodule Keyword do
+      iex> Keyword.validate([three: 3, four: 4], [one: 1, two: 2])
+      {:error, {:invalid_key, :three}}
+
+  Passing an invalid value in either argument raises
+
+      iex> Keyword.validate(%{three: 3}, [one: 1, two: 2])
+      ** (ArgumentError) expected a keyword list with keys [:one, :two], got: %{three: 3}
+
+      iex> Keyword.validate([:three], [one: 1, two: 2])
+      ** (ArgumentError) expected a keyword list with keys [:one, :two], got: [:three]
+
+      iex> Keyword.validate([three: 3], [:three, 3, :two])
+      ** (ArgumentError) expected the second argument to be a list of atoms or tuples, got item: 3
```
Ok, I'll move these to unit tests

polvalente commented:
@wojtekmach For some reason Github is not letting me "re-request" your review. So, I've refactored the code and it's ready for another round :)

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,142 @@ defmodule Keyword do
+      {:error, invalid_keys} ->
+        keys =
+          for value <- values,
+              do: Kernel.if(is_atom(value), do: value, else: Kernel.elem(value, 0))
```
```suggestion
              do: if(is_atom(value), do: value, else: elem(value, 0))
```

polvalente replied:
copy-paste slipped :grimacing:

josevalim commented on lib/elixir/test/elixir/keyword_test.exs:
```diff
@@ -205,4 +205,24 @@ defmodule KeywordTest do
     end
   end
+
+  test "validate/2 raises on invalid arguments" do
+    assert_raise ArgumentError,
+                 "expected a keyword list with keys [:one, :two], got: %{three: 3}",
+                 fn ->
+                   Keyword.validate(%{three: 3}, one: 1, two: 2)
+                 end
+
+    assert_raise ArgumentError,
+                 "expected a keyword list with keys [:one, :two], got: [:three]",
+                 fn ->
+                   Keyword.validate([:three], one: 1, two: 2)
+                 end
+
+    assert_raise ArgumentError,
+                 "expected the second argument to be a list of atoms or tuples, got item: 3",
+                 fn ->
+                   Keyword.validate([three: 3], [:three, 3, :two])
+                 end
+  end
```
```suggestion
  test "validate/2 raises on invalid arguments" do
    assert_raise ArgumentError,
                 "expected a keyword list as first argument, got invalid entry: :three",
                 fn -> Keyword.validate([:three], one: 1, two: 2) end

    assert_raise ArgumentError,
                 "expected the second argument to be a list of atoms or tuples, got: 3",
                 fn -> Keyword.validate([three: 3], [:three, 3, :two]) end
  end
```

josevalim commented:
@polvalente I have reverted to the original implementation since it was optimized to avoid lists traversals and kept the rest. Next, can you also add Map.validate? I am wondering what would be the best way to implement it... :)

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,142 @@ defmodule Keyword do
+      {:error, invalid_keys} ->
+        keys =
+          for value <- values,
+              do: if(is_atom(value), do: value, else: Kernel.elem(value, 0))
```
```suggestion
              do: if(is_atom(value), do: value, else: elem(value, 0)
```

polvalente commented:
> @polvalente I have reverted to the original implementation since it was optimized to avoid lists traversals and kept the rest. Next, can you also add Map.validate? I am wondering what would be the best way to implement it... :)

@josevalim Sure, no problem!

I had changed the implementation because it was easier than refactoring yours for accumulating multiple invalid keys.

Should Map.validate come in another PR? I'm thinking that perhaps there is a way to use Keyword.validate underneath it, if the stdlib itself doesn't complain about it the way it complained about Enum. I'm not sure about the cost of converting from map to keyword back to map.

Another approach I have in mind is to look at how `@enforce_keys` is implemented for structs, since it's kind of a similar behavior, as well as looking into `:maps` to see if any functions there could help.

WDYT?

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,142 @@ defmodule Keyword do
+
+  The second argument must be a list of atoms, specifying
+  a given key, or tuples specifying a key and a default value.
+
+
+  It returns `{:ok, keyword}` in case the keyword list only has
+  the given keys with returns the validated keyword with default
+  values applied. Otherwise it returns `{:error, invalid_keys}`
+  with invalid keys.
```
```suggestion

  If the keyword list has only the given keys, it returns
  `{:ok, keyword}` with default values applied. Otherwise it
  returns `{:error, invalid_keys}` with invalid keys.
```

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,142 @@ defmodule Keyword do
+      iex> Enum.sort(result)
+      [one: 1, two: 2]
+
+  Passing an unknown key returns an error:
```
```suggestion
  Passing unknown keys returns an error:
```

wojtekmach commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,142 @@ defmodule Keyword do
+      iex> Keyword.validate!([one: 1], [:one, two: 2]) |> Enum.sort()
+      [one: 1, two: 2]
+
+  Passing an unknown key raises:
```
```suggestion
  Passing unknown keys raises an error:
```

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -186,6 +186,140 @@ defmodule Keyword do
+      {:error, invalid_keys} ->
+        keys =
+          for value <- values,
+              do: if(is_atom(value), do: value, else: elem(value, 0)
```
```suggestion
              do: if(is_atom(value), do: value, else: elem(value, 0))
```

josevalim commented:
I am glad to sit on Map.validate until someone submits it or comes up with a use case. I will be merging this once CI passes then!

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
